pkg/timeout/backoff.go

```diff
@@ -0,0 +1,148 @@
+
+// NewBackoff returns an instance of Backoff with the specified stopCh and logger, instantiated
+// and ready to track go routines.
+func NewBackoff(
+	stopCh <-chan struct{},
+	logger *zap.SugaredLogger,
+) *Backoff {
+	return &Backoff{
+		timer:    NewTimer(stopCh, logger),
+		attempts: make(map[string]Attempts),
+		j:        rand.Intn,
+		now:      time.Now,
+		logger:   logger,
+	}
+}
+
+// Release will remove keys tracking the specified runObj.
+func (b *Backoff) Release(runObj StatusKey) {
+	b.attemptsMut.Lock()
+	defer b.attemptsMut.Unlock()
+	delete(b.attempts, runObj.GetRunKey())
+}
+
+// SetTimeoutCallback will set the function to be called when a timeout has occurred.
+func (b *Backoff) SetTimeoutCallback(f func(interface{})) {
+	b.timeoutCallback = f
+}
+
+// Get records the number of times it has seen a TaskRun and calculates an
+// appropriate backoff deadline based on that count. Only one backoff per TaskRun
+// may be active at any moment. Requests for a new backoff in the face of an
+// existing one will be ignored and details of the existing backoff will be returned
+// instead. Further, if a calculated backoff time is after the timeout of the TaskRun
+// then the time of the timeout will be returned instead.
+//
+// Returned values are a backoff struct containing a NumAttempts field with the
+// number of attempts performed for this TaskRun and a NextAttempt field
+// describing the time at which the next attempt should be performed.
+// Additionally a boolean is returned indicating whether a backoff for the
+// TaskRun is already in progress.
+func (b *Backoff) Get(tr *v1beta1.TaskRun) (a Attempts, inProgress bool) {
+	b.attemptsMut.Lock()
+	defer b.attemptsMut.Unlock()
+	a = b.attempts[tr.GetRunKey()]
+	if b.now().Before(a.NextAttempt) {
+		inProgress = true
+		return
+	}
+	a.NumAttempts++
+	a.NextAttempt = b.now().Add(GetExponentialBackoffWithJitter(a.NumAttempts, b.j))
+	duration := timeoutFromSpec((tr.Spec.Timeout))
+	timeoutDeadline := tr.Status.StartTime.Time.Add(duration)
+	if timeoutDeadline.Before(a.NextAttempt) {
+		a.NextAttempt = timeoutDeadline
+	}
+	b.attempts[tr.GetRunKey()] = a
+	return
+}
+
+// GetExponentialBackoffWithJitter will return a number which is 2 to the power
+// of count, but with a jittered value obtained via jf.
+func GetExponentialBackoffWithJitter(count uint, jf jitterFunc) time.Duration {
+	exp := float64(count)
+	if exp > maxBackoffExponent {
+		exp = maxBackoffExponent
+	}
+	seconds := int(math.Exp2(exp))
+	jittered := 1 + jf(seconds)
+	if jittered > maxBackoffSeconds {
+		jittered = maxBackoffSeconds
+	}
+	return time.Duration(jittered) * time.Second
+}
+
+// SetTimer creates a blocking function to wait for
+// 1. Stop signal, 2. completion or 3. a given Duration to elapse.
+func (b *Backoff) SetTimer(runObj StatusKey, d time.Duration) {
+	if b.timeoutCallback == nil {
+		b.logger.Errorf("attempted to set a timer for %q but no callback has been assigned", runObj)
```

Comment:
Sounds very odd (at least to me 😉), commenting out this `logger` from here fixes race condition. This function invocation within [handlePodCreationError](https://github.com/tektoncd/pipeline/pull/3031/files?diff=split&w=1#diff-14481606a469dcd50f6cb9f5540e3b06R467) and the following [SetCondition](https://github.com/tektoncd/pipeline/pull/3031/files?diff=split&w=1#diff-14481606a469dcd50f6cb9f5540e3b06R470) are somehow causing this.

Reply:
```
WARNING: DATA RACE
Read at 0x00c000c3c198 by goroutine 103:
...
go.uber.org/zap.(*SugaredLogger).log()
...

Previous write at 0x00c000c3c198 by goroutine 79:
  knative.dev/pkg/apis/duck/v1beta1.(*Status).SetConditions()
...
```

Reply:
Ah, well spotted!  It looks like swapping out the `%q` for a `%p` in this log statement also removes the error.  Using `%q` results in a string-formatted struct being printed, which probably walks the entire object and accesses all the fields.  My hunch is that `%q` is not threadsafe in that case?

Reply:
nice catch, yup it does walk the entire object. The [next](https://github.com/tektoncd/pipeline/pull/3031/files#diff-fc314ed1e1edf4a1e359bff0f4f2c47aR145) log statement (using `GetRunKey()` on `runObj`) looks safe and just prints the address of `taskRun` object instead.